src/cluster.c

```diff
@@ -1030,9 +1030,11 @@ clusterNode *getNodeByQuery(client *c, struct redisCommand *cmd, robj **argv, in
         mc.cmd = cmd;
     }
 
-    int is_pubsubshard = cmd->proc == ssubscribeCommand ||
-                         cmd->proc == sunsubscribeCommand ||
-                         cmd->proc == spublishCommand;
+    uint64_t cmd_flags = getCommandFlags(c);
+    
+    /* Only valid for sharded pubsub as regular pubsub can operate on any node and bypasses this layer. */
+    int is_pubsubshard = (cmd_flags & CMD_PUBSUB) ||
+                           (c->cmd->proc == execCommand && (c->mstate.cmd_flags & CMD_PUBSUB));
```

Comment:
There are two issues:
1. Whenever a transaction contains shard pubsub commands, the entire transaction is considered as `is_pubsubshard`, which is too broad of a scope. Although it may not have any practical impact at the moment, there is a potential risk. Perhaps we should rename the variable to `include_pubsubshard`.
2. Non-shard pubsub commands, such as `publish`/`subscribe`, are also marked as `CMD_PUBSUB`, which leads to incorrect determination of `publish`/`subscribe` in transaction as `is_pubsubshard`. Although it may not have any practical impact currently, there is a potential risk.

Reply:
Thanks for the feedback @soloestoy.

Agree for point 1, will update it.
I did think about point hence the comment. Explicit command matching across multi command would be slightly painful and introducing a new command type felt unnecessary at this point. Do you think it's reasonable to continue with the current implementation.